hylafaxplus: fix with new `libtiff_4_5`

Yarny0 opened the pull request:
## Description of changes

With the [update to libtiff 4.6.0](https://github.com/NixOS/nixpkgs/pull/261791), the library [dropped many tiff processing utility executables](https://github.com/NixOS/nixpkgs/pull/261791#issuecomment-1771689841), like `tiff2ps` and `fax2ps`.  These executables are needed at least by `hylafaxplus` (possibly also by other packages).

The pull request at hand introduces `libtiff_4_5` to nixpkgs in order to ~~avoid breaking the `hylafaxplus` package when https://github.com/NixOS/nixpkgs/pull/263535 gets merged~~ fix `hylafaxplus`.

libtiff 4.6.0 also [fixed two CVEs](https://github.com/NixOS/nixpkgs/pull/261791#issuecomment-1768177356).  `libtiff_4_5` includes the corresponding patches from the libtiff development repository.


## Long-term prospect

[According to HylaFAX+ developer Lee Howard](https://sourceforge.net/p/hylafax/mailman/message/38259441/), it might take months before one of these options (or a combination thereof) might be implemented:

*   a new package `tifftools` with the utility executables emerges
*   the HylaFAX+ source repository incorporates the source of the required executables and builds them by itself
*   HylaFAX+ gets rewritten so as to not use these programs
*   libtiff reincorporates the programs, maybe after fixing them

He already initiated [a discussion on the libtiff mailing list](https://www.asmail.be/msg0054915176.html).  In the meantime, [he recommends](https://sourceforge.net/p/hylafax/mailman/message/39730813/) for distributions with rolling releases to do what they usually do when such a situation arises.


## Notifications

*   libitff maintainer @alyssais : Should I also mention you as maintainer of `libtiff_4_5` ?
*   @trofi @risicle due to [your discussion about the removal of the executables](https://github.com/NixOS/nixpkgs/pull/261791#issuecomment-1771689841)


## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/tes>
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkg>
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/bl>
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

I tested the `hylafaxplus` package (with asterisk and a couple of iaxmodem instances) to send and receive some faxes, and it still works as before -- which is to be expected, as the only change from HylaFAX+' perspective are a few patches for libtiff 4.5.1.

## Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>hylafaxplus</li>
    <li>libtiff_4_5</li>
    <li>libtiff_4_5.bin</li>
    <li>libtiff_4_5.dev</li>
    <li>libtiff_4_5.dev_private</li>
    <li>libtiff_4_5.doc</li>
    <li>libtiff_4_5.man</li>
  </ul>
</details>

Yarny0 committed:
libtiff: introduce libtiff_4_5

With the update to libtiff 4.6 in
0a74a54ac2600656cd0b640d5ea1c8efb5c35d68 ,
many tiff-processing utility executables got dropped:

http://www.simplesystems.org/libtiff/releases/v4.6.0.html

Some of these executables can still be "restored" with
the configure switch `--enable-tools-unsupported`,
but unfortunatelly,
at least hylafaxplus (maybe more packages) relies on
utilities that even cannot be restored with this switch.

The commit at hand reintroduces the old libtiff
version 4.5.1 as `libtiff_4_5` into nixpkgs.
It restores the old build recipe with the following changes:

* passthru.updateScript is dropped as it is of no use here
* passthru.tests is dropped as it only contains
  packages that now build with the new libtiff version
* patches are applied for the two CVEs that are fixed in 4.6.0

As libtiff 4.5 is no longer supported by libtiff developers,
new vulnerabilities will likely go unnoticed
unless they also affect the current version.
To not disable hydra builds, we don't add
`knownVulnerabilities` *for now*, but add comments to alert
updaters of the current libtiff version so patches can
be backported or the situation be reevaluated as a whole.

Yarny0 committed:
hylafaxplus: fix by downgrading libtiff to 4.5.1

libtiff 4.6.0 (updated with
0a74a54ac2600656cd0b640d5ea1c8efb5c35d68)
drops many tiff-processing utility executables, like
tiff2ps and fax2ps, that are needed by hylafaxplus.
As a result, the build fails with

> FATAL ERROR: /nix/store/5i80biwsgjb4yhp4z81nsk3kp5rivd8f-libtiff-4.6.0-bin/bin/tiff2ps does not exist or is not an executable program!

see https://hydra.nixos.org/build/239580858 .

The commit at hand fixes the package by using libtiff_4_5
which got introduced in the previous commit.

Note on long-term prospect for this situation:
According to hylafaxplus developer Lee Howard,
it might take months before one of these options
(or a combination thereof) might be implemented:

* a new package "tifftools" with the utility executables emerges
* The hylafaxplus package incorporates the source
  of these executables and builds them by itself
* hylafaxplus is rewritten so as to not use these programs
* libtiff reincorporates the programs, maybe after fixing them

Cf. https://sourceforge.net/p/hylafax/mailman/message/38259441/

Another reference: a discussion on the libtiff mailing list:
https://www.asmail.be/msg0054915176.html

trofi commented on pkgs/top-level/all-packages.nix:
```diff
@@ -9296,7 +9296,7 @@ with pkgs;
   };
 
-  hylafaxplus = callPackage ../servers/hylafaxplus { };
+  hylafaxplus = callPackage ../servers/hylafaxplus { libtiff = libtiff_4_5; };
```
I suggest adding inline comment here why you need pinned version: just mention needed utilities here.

Otherwise looks great!

Yarny0 replied:
Done.

alyssais commented:
> libitff maintainer @alyssais : Should I also mention you as maintainer of `libtiff_4_5` ?

No, in fact I should step down as maintainer of libtiff.  I didn't realise what I was getting into and haven't really kept up.

risicle commented:
Could you point to how you identified the CVEs' corresponding patches? It's not obvious to me, and it would be nice to have them in 23.05.

It does concern me a bit re-introducing these with no `knownVulnerabilities` marker. Upstream is dropping them AFAIK because it considers them unfixable. Their removal also means that libtiff upstream aren't going to issue any more CVEs for the tools (who knows if they have already received more such reports?). So it feels a _little_ like solving a house-fire by unplugging the smoke alarm.

Yarny0 commented:
> Could you point to how you identified the CVEs' corresponding patches?

Honestly, these are just the CVEs listed in [that comment](https://github.com/NixOS/nixpkgs/pull/261791#issuecomment-1768177356).  There might be more.

Currently, HylaFAX+ author Lee Howard [expects vulnerabilities in dropped executables to be irrelevant to HylaFAX+](https://www.asmail.be/msg0054917226.html).  However, that certainly can not account for vulnerabilities discovered in the future.

I updated the branch to include a corresponding `knownVulnerabilities` entry.  That should allow HylaFAX+ (and other possible `libtiff_4_5`) users to make an informed decision about using it.

risicle commented:
> Honestly, these are just the CVEs listed in https://github.com/NixOS/nixpkgs/pull/261791#issuecomment-1768177356. There might be more.

But I can't see how those CVE numbers lead to the patches you're pulling.

Yarny0 commented:
### CVEs

> But I can't see how those CVE numbers lead to the patches you're pulling.

The search was complicated and I didn't make notes (and thereby accidentally switched CVE numbers, see below).  But I was able to reconstruct what I was doing:

*   From looking at https://www.cvedetails.com/vulnerability-list/vendor_id-2224/Libtiff.htm , these are the only two CVEs (currently) known against libtiff 4.5.1.

For each CVE link in https://github.com/NixOS/nixpkgs/pull/261791#issuecomment-1768177356 I did this (I'm sorry if this is now too much details -- but better than too few):

*   follow the CVE link
*   follow the bugzilla.redhat.com link on the CVE info page
*   the bug has exactly one entry in its list of duplicate bug entries: follow the link, to another bug bugzilla.redhat.com
*   follow the link in comment number 3 to an issue/ticket in libtiff's gitlab project
*   there is a list of merge requests, with only one accepted merge request: follow the link to that one
*   the merge request has one commit

I have added those commits as patches to `libtiff_4_5`.

Thanks to your question and my reconstructing this path of links, I discovered that I have accidentally switched CVE numbers in my `fetchpatch` invocations.  I have fixed this and pushed a new version.


### NixOS 23.05

> Could you point to how you identified the CVEs' corresponding patches? It's not obvious to me, and it would be nice to have them in 23.05.

I will prepare a pull request for `staging-23.05` and put it in draft mode, so we can first sort out remaining issues here. I hope that's OK.


### Hydra builds

From what I see when looking at other packages with `knownVulnerabilities`, Hydra apparently won't build them.  This is a bit unfortunate as users of HylaFAX+ (like myself) will need to build `libtiff_4_5` and `hylafaxplus` locally, with corresponding CPU usage.  It also prevents me from following hydra build reports to discover problems with the package early (I'm using the stable channels for my machines).

Do you see any way to include a warning about possible vulnerabilities without stopping hydra builds?

risicle commented:
> For each CVE link in https://github.com/NixOS/nixpkgs/pull/261791#issuecomment-1768177356 I did this...

Awesome thanks, I had missed this. I should have noticed that the CVEs are actually named in the issues themselves, I could have searched for them.


`knownVulnerabilities` is indeed a rather blunt tool, but we don't have any other mechanism. It also stops `nixpkgs-review` from testing a package which can lead to further bitrot.

Arguably we're not proposing adding `knownVulnerabilities` to a patched libtiff 4.5 in the stable branch, which is exactly the same piece of software, so why would we add it here? I think I'm worried about the possibility of this remaining in this state long-term.

Yarny0 commented:
> Awesome thanks, I had missed this.

I think the way from the CVE to the patch is really non-obvious, and the most non-obvious part is probably the duplicate bugzilla bug entry.  I have added links to those as comment in the build recipe (and in https://github.com/NixOS/nixpkgs/pull/264613) to make it more easy for others who want to follow that path without searching for the discussion on the pull request at hand.

> Arguably we're not proposing adding knownVulnerabilities to a patched libtiff 4.5 in the stable branch, which is exactly the same piece of software, so why would we add it here? I think I'm worried about the possibility of this remaining in this state long-term.

How about this:

We add, but comment out, the `knownVulnerabilities` attribute for now, and stick a prominent "XXX" comment to it to alert updaters.  Also, we add a corresponding comment in `default.nix` to ask updaters to watch out for security patches and consider updating `4.5.nix`.  I'm a bit reluctant to polluting `default.nix` with such tasks, but when the currently supported version receives an update, that's probably the best moment to look for patches that might be worth being applied to 4.5 as well.

I have updates the branch accordingly, and the way the comment is placed in `default.nix` should make it show up in a diff even on updates generated by r-ryantm bot.  While it is not foolproof, it should minimize the probability of `4.5.nix` to go unnoticed for years.  I also updated the commit message with a short explanation.

(Of course, I also try to keep an eye on the things I maintain, but I like it to be designed that anyone could pick this up with minimal friktion)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2883

Yarny0 commented:
Now that staging-next got merged in master, I updated the PR text and added the result of nixpkgs-review.

This PR now qualifies for the ZHF-label #265948 .

It would be nice to fix HylaFAX in time for NixOS 23.11.
